helper/fairshare/jobmanager.go

```diff
@@ -179,10 +177,94 @@ func (j *JobManager) getNextJob() Job {
 	}
 
 	if j.queues[queueID].Len() == 0 {
-		j.removeLastQueueAccessed()
+		// we remove the empty queue, but we don't remove the worker count
+		// in case we are still working on previous jobs from this queue.
+		// worker count cleanup is handled in j.decrementWorkerCount
+		delete(j.queues, queueID)
+	}
+
+	return out.(Job), queueID
+}
+
+// returns the next queue to assign work from, and a bool if there is a queue
+// that can have a worker assigned.
+// the intent is to avoid over-allocating work from specific queues, as
+// outlined in RFC VLT-145
+// note: this must be called with j.l held
+func (j *JobManager) getNextQueue() (string, bool) {
+	var nextQueue string
+	var canAssignWorker bool
+
+	queueIDsByIncreasingWorkers := j.sortByNumWorkers()
+	for _, queueID := range queueIDsByIncreasingWorkers {
+		if !j.queueWorkersSaturated(queueID) {
+			nextQueue = queueID
+			canAssignWorker = true
+			break
+		}
 	}
 
-	return out.(Job)
+	return nextQueue, canAssignWorker
+}
+
+// returns true if there are already too many workers on this queue
+// note: this must be called with j.l held (at least for read).
+// note: we may want to eventually factor in queue length relative to num queues
+func (j *JobManager) queueWorkersSaturated(queueID string) bool {
+	numActiveQueues := float64(len(j.queues))
+	numTotalWorkers := float64(j.workerPool.numWorkers)
+	maxWorkersPerQueue := math.Ceil(0.9 * numTotalWorkers / numActiveQueues)
+
+	numWorkersPerQueue := j.workerCount
+
+	return numWorkersPerQueue[queueID] >= int(maxWorkersPerQueue)
+}
+
+// sortByNumWorkers returns queueIDs in order of increasing number of workers
+// note: this must be called with j.l held
+func (j *JobManager) sortByNumWorkers() []string {
+	out := make([]string, 0, len(j.queues))
+	for queueID := range j.queues {
+		out = append(out, queueID)
+	}
+
+	workersPerQueue := j.workerCount
+
+	sort.Slice(out, func(i, j int) bool {
+		// TODO should we be explicitly breaking times with the queueID, or
```

Comment:
i'd love some thoughts on this

Reply:
I think you're right, randomness here wouldn't necessarily give us anything as it wouldn't decrease the stress on any of the workers (since this is just a tiebreaker).

Reply:
You can't use randomness purely, as some sorts won't ever complete if the tests aren't consistent.

Reply:
the randomness i was discussing is from the order from the map iter, so whatever is in the sort less function would be consistent

basically, should we leave the tiebreaker in or not?